nn/conformer.py

```diff
@@ -0,0 +1,215 @@
+      batch_norm_other_opts = {}
+    self.batch_norm = nn.BatchNorm(
+      epsilon=batch_norm_eps, momentum=batch_norm_momentum, update_sample_only_in_training=True,
+      delay_sample_update=True, **batch_norm_other_opts)
+
+  @staticmethod
+  def _glu(v: LayerRef):
+    a, b = nn.split(v, axis='F')
+    return a * nn.sigmoid(b)
+
+  def forward(self, inp: LayerRef) -> LayerRef:
+    x_conv1 = self.positionwise_conv1(inp)
+    x_act = self._glu(x_conv1)
+    x_depthwise_conv = self.depthwise_conv(x_act)
+    x_bn = self.batch_norm(x_depthwise_conv)
+    x_swish = nn.swish(x_bn)
+    x_conv2 = self.positionwise_conv2(x_swish)
+    return x_conv2
+
+
+class _ConformerConvSubsampleLayer(nn.Module):
+  """
+  Conv 2D block with optional max-pooling
+  """
+
+  def __init__(self, filter_sizes: List[Tuple[int, ...]], pool_sizes: Union[List[Tuple[int, ...]], None],
+      channel_sizes: List[int], l2: float = 0.0, dropout: float = 0.3, activation: str = 'relu',
+      padding: str = 'same'):
+    """
+    :param filter_sizes:
+    :param pool_sizes:
+    :param channel_sizes:
+    :param l2:
+    :param dropout:
+    :param activation:
+    :param padding:
+    """
+    super().__init__()
+
+    self.dropout = dropout
+    self.pool_sizes = pool_sizes
+
+    self.conv_layers = nn.ModuleList()
+    for filter_size, channel_size in zip(filter_sizes, channel_sizes):
+      self.conv_layers.append(
+        nn.Conv(l2=l2, activation=activation, filter_size=filter_size, n_out=channel_size, padding=padding))
+
+  def forward(self, inp: LayerRef) -> LayerRef:
+    x = nn.split_dims(inp, axis='F', dims=(-1, 1))
+    for i, conv_layer in enumerate(self.conv_layers):
+      x = conv_layer(x)
+      if self.pool_sizes and i < len(self.pool_sizes):
+        x = nn.pool(x, pool_size=self.pool_sizes[i], padding='same', mode='max')
+      if self.dropout:
+        x = nn.dropout(x, dropout=self.dropout)
+    out = nn.merge_dims(x, axes='static')
+    return out
+
+
+class ConformerEncoderLayer(nn.Module):
+  """
+  Represents a conformer block
+  """
+
+  def __init__(self, conv_kernel_size: int, activation_ff, dim_ff: int, dropout: float, att_dropout: float,
+      enc_key_dim: int, num_heads: int, l2: float):
+    """
+    :param conv_kernel_size:
+    :param activation_ff:
+    :param ff_dim:
+    :param dropout:
+    :param att_dropout:
+    :param enc_key_dim:
+    :param num_heads:
+    :param l2:
+    """
+    super().__init__()
+
+    self.dropout = dropout
+
+    self.ffn1 = _PositionwiseFeedForward(
+      dim_model=enc_key_dim, dim_ff=dim_ff, dropout=dropout, activation=activation_ff, l2=l2)
+
+    self.ffn2 = _PositionwiseFeedForward(
+      dim_model=enc_key_dim, dim_ff=dim_ff, dropout=dropout, activation=activation_ff, l2=l2)
+
+    self.conv_module = _ConformerConvBlock(dim_model=enc_key_dim, kernel_size=conv_kernel_size)
+
+    self.mhsa_module = MultiheadAttention(enc_key_dim, num_heads, dropout=att_dropout)  # TODO: to be implemented
```

Comment:
Same here on the attrib name. Maybe just `mhsa`. Or maybe better `self_att` (I don't think the multi-head aspect is so relevant to put it into the attrib name).